primitives/src/balances.rs

```diff
@@ -56,6 +57,91 @@ pub struct BalanceEntry<BlockNumber> {
 	pub last_update: BlockNumber,
 }
 
+impl<BlockNumber> BalanceEntry<BlockNumber>
+where
+	BlockNumber: AtLeast32Bit,
+{
+	pub fn new(principal: BalanceType, last_update: BlockNumber) -> Self {
+		Self { principal, last_update }
+	}
+
+	/// Applies the demurrage and returns an updated BalanceEntry.
+	///
+	/// The following formula is applied to the principal:
+	///    updated_principal = old_principal * e^(-demurrage_per_block * elapsed_blocks)
+	///
+	/// **Note**: This function will be used at every single transaction that is paid with community
+	/// currency. It is important that it is as efficient as possible, but also that it is bullet-
+	/// proof (no-hidden panics!!).
+	pub fn apply_demurrage(
+		self,
+		demurrage_per_block: Demurrage,
+		current_block: BlockNumber,
+	) -> Result<BalanceEntry<BlockNumber>, DemurrageError> {
+		if demurrage_per_block < 0 {
+			return Err(DemurrageError::DemurrageMustBePositive)
+		}
+
+		if self.last_update == current_block {
+			// Nothing to be done, as no time elapsed.
+			return Ok(self)
+		}
+
+		if self.principal.eq(&0i16) {
```

Comment:
Is it worth doing this check every time?
E.g. what's more often the case: `principal ==` 0 or `principal !=0`? If it's the latter, isn't it a waste of time?

Reply:
`principal == 0` it should not be more often, but at this level we don't know. The overflow bug we had, was on a `BalanceEntry::default()`, which was created when an account did not exist, and then the demurrage was applied since genesis. Of course the overflow error is fixed now, but it is just so much cheaper to make this check here more often than to perform the entire demurrage computation in vain.